### PR TITLE
feat: add oauth api support gitlab(v3 v4) and github v3 version

### PR DESCRIPTION
**Pull request by wzqnls**
**What this PR does / why we need it**:

Add your description

**Which issue(s) this PR is related to** *(optional, link to 3rd issue(s))*:

Fixes #

Reference to #

**Special notes for your reviewer**:

/cc @your-reviewer

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/api/v1/handler/oauth.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+	if err != nil {
+		return err
+	}
+	username, server, err := client.GetUserInfo(token)
+	if err != nil {
+		return err
+	}
+	consoleWebServer := os.GetStringEnv(options.ConsoleWebEndpoint, "")
+	redirectURL := fmt.Sprintf(
+		"%s/devops/workspace/add?type=gitlab&code=%s&state=%s&token=%s&username=%s&server=%s",
```
Comment (zhujian7) [inline pkg/api/v1/handler/oauth.go]: what does this `devops/workspace/add` represent? Is it general ?

File: pkg/util/oauth/provider.go [hunk 2]
```diff
@@ -0,0 +1,223 @@
+package oauth
```
Comment (zhujian7) [inline pkg/util/oauth/provider.go]: I think this file is not utils for common invoking, and should not put it in `util` folder. @wzqnls  WDYT?

File: pkg/util/oauth/provider.go [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+		if err != nil {
+			return "", "", errors.ErrorUnknownInternal.Error(err)
+		}
+		return *user.Login, GithubURL, nil
+	default:
+		return "", "", errors.ErrorUnsupported.Error("scm type", oc.scm)
+	}
+}
+
+// DetectAPIVersion to deect api version of gitlab v3 or v4
```
Comment (zhujian7) [inline pkg/util/oauth/provider.go]: detect?

**Comment by zhujian7** [comment]
@wzqnls Could you please add some docs to [here](https://github.com/caicloud/cyclone/blob/master/docs/api/v1/api.md)?

**Comment by zhujian7** [comment]
@supereagle PTAL

File: pkg/util/oauth/provider.go (hunk 2 above)
Comment (supereagle) [reply pkg/util/oauth/provider.go]: I think it would be better to put in `pkg/scm/provider`.

**Comment by supereagle** [comment]
@wzqnls Please fix comments.

**Comment by supereagle** [comment]
out of date

**Pull request closed without merging**